patches/api/0453-Brigadier-based-command-API.patch

```diff
@@ -0,0 +1,1090 @@
++interface VanillaArgumentProvider {
++
++    ArgumentType<SelectorArgumentResolver<Entity>> entity();
++
++    ArgumentType<SelectorArgumentResolver<Player>> player();
++
++    ArgumentType<SelectorArgumentResolver<List<Entity>>> entities();
++
++    ArgumentType<SelectorArgumentResolver<List<Player>>> players();
++
++    ArgumentType<ArgumentResolver<Collection<PlayerProfile>>> playerProfiles();
++
++    ArgumentType<ArgumentResolver<BlockPosition>> blockPos();
++
++    ArgumentType<BlockState> blockState();
++
++    ArgumentType<ItemStack> itemStack();
++
++    ArgumentType<NamedTextColor> namedColor();
++
++    ArgumentType<Component> component();
++
++    ArgumentType<MessageArgumentResponse> signedMessage();
++
++    ArgumentType<DisplaySlot> scoreboardDisplaySlot();
++
++    ArgumentType<NamespacedKey> namespacedKey();
++
++    // include both key types as we are slowly moving to use adventure's key
++    ArgumentType<Key> key();
++
++    ArgumentType<Range<Integer>> integerRange();
++
++    ArgumentType<Range<Double>> doubleRange();
++
++    ArgumentType<World> world();
++    
++    ArgumentType<GameMode> gameMode();
++
++    ArgumentType<HeightMap> heightMap();
++
++    ArgumentType<UUID> uuid();
++
++    void validateNativeType(ArgumentType<?> type);
++}
+diff --git a/src/main/java/io/papermc/paper/command/brigadier/argument/VanillaArgumentTypes.java b/src/main/java/io/papermc/paper/command/brigadier/argument/VanillaArgumentTypes.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..18449f4927a2d91f24bcc3e1f83b982044b6ad16
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/command/brigadier/argument/VanillaArgumentTypes.java
+@@ -0,0 +1,256 @@
++package io.papermc.paper.command.brigadier.argument;
++
++import com.destroystokyo.paper.profile.PlayerProfile;
++import com.google.common.collect.Range;
++import com.mojang.brigadier.arguments.ArgumentType;
++import com.mojang.brigadier.context.CommandContext;
++import io.papermc.paper.command.brigadier.CommandSourceStack;
++import io.papermc.paper.math.BlockPosition;
++import net.kyori.adventure.key.Key;
++import net.kyori.adventure.text.Component;
++import net.kyori.adventure.text.format.NamedTextColor;
++import org.bukkit.GameMode;
++import org.bukkit.HeightMap;
++import org.bukkit.NamespacedKey;
++import org.bukkit.World;
++import org.bukkit.block.BlockState;
++import org.bukkit.entity.Entity;
++import org.bukkit.entity.Player;
++import org.bukkit.inventory.ItemStack;
++import org.bukkit.scoreboard.DisplaySlot;
++import org.jetbrains.annotations.ApiStatus;
++import org.jetbrains.annotations.NotNull;
++
++import java.util.Collection;
++import java.util.List;
++import java.util.ServiceLoader;
++import java.util.UUID;
++
++/**
++ * Vanilla argument registry
++ * <p>
++ * These provide rich argument parsing on the client and
++ * may also provide additional signing context.
++ */
++@ApiStatus.Experimental
++public final class VanillaArgumentTypes {
```

Comment:
`ArgumentTypes` or `CommandArgumentTypes` since it's kinda obvious these are the default ones provided by the server. + add a link to where you can register/implement your own in docs

Reply:
I made this distinction in the case that other plugins would want to provide their own argument types, but if people disagree with that I can rename this.

Reply:
We don't do this with anything else that is registerable as well, so this just breaks conistency

Reply:
👍

Reply:
But to note, the only place that this is done IIRC is the mob goal api.

Reply:
Oh yeah yuck, time to fire mini